### PR TITLE
Improve address decoding errors

### LINKED ISSUE #21741: Error messages for invalid address
Can we add more information in error messages for invalid address especially the errors mentioned in PR: https://github.com/bitcoin/bitcoin/pull/20832/

Example: 

Change https://github.com/bitcoin/bitcoin/blob/f385ad765174afb02e60900581612a19c143cf83/src/key_io.cpp#L103

To

```C++
error_str = "Invalid prefix for Bech32 address. Valid Bech32 address starts with `bc1` (mainnet) or `tb1` (testnet)";
```

Context: https://github.com/bitcoin-core/gui/pull/280#issuecomment-820957002

### PR DESCRIPTION
**Pull request by aureleoules**
Attempt to fix #21741.

### CODE REVIEW

File: src/key_io.cpp [hunk 1]
```diff
@@ -107,17 +107,17 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
                 std::equal(script_prefix.begin(), script_prefix.end(), data.begin())) ||
             (data.size() >= pubkey_prefix.size() &&
                 std::equal(pubkey_prefix.begin(), pubkey_prefix.end(), data.begin()))) {
-            error_str = "Invalid length for Base58 address";
+            error_str = "Invalid length for Base58 address (P2PKH or P2SH)";
         } else {
-            error_str = "Invalid prefix for Base58-encoded address";
+            error_str = strprintf("Invalid prefix for Base58-encoded address on %s. This may be an address for a different network. Valid address starts with `1` or `3` (mainnet) and `m` or `n` or `2` (testnet).", params.NetworkIDString());
```
Comment (luke-jr) [inline src/key_io.cpp:112]:
```suggestion
            error_str = strprintf("Invalid or unsupported Base58-encoded address.", params.NetworkIDString());
```

File: src/key_io.cpp [hunk 2]
```diff
@@ -107,17 +107,17 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
...
         }
         return CNoDestination();
     } else if (!is_bech32) {
         // Try Base58 decoding without the checksum, using a much larger max length
         if (!DecodeBase58(str, data, 100)) {
-            error_str = "Not a valid Bech32 or Base58 encoding";
+            error_str = "Not a valid Segwit (Bech32) or Base58 encoding. This may be an address for a different network.";
```
Comment (luke-jr) [inline src/key_io.cpp:118]:
```suggestion
            error_str = "Invalid or unsupported Segwit (Bech32) or Base58 encoding.";
```

File: src/key_io.cpp [hunk 3]
```diff
@@ -127,7 +127,8 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
     if ((dec.encoding == bech32::Encoding::BECH32 || dec.encoding == bech32::Encoding::BECH32M) && dec.data.size() > 0) {
         // Bech32 decoding
         if (dec.hrp != params.Bech32HRP()) {
-            error_str = "Invalid prefix for Bech32 address";
+            error_str = strprintf("Invalid prefix for Segwit (Bech32) address on %s (expected %s, got %s). This may be an address for a different network.",
```
Comment (luke-jr) [inline src/key_io.cpp:130]:
```suggestion
            error_str = strprintf("Invalid or unsupported prefix for Segwit (Bech32) address on %s (expected %s, got %s).",
```

**Review by luke-jr — CHANGES_REQUESTED** [review]
NACK on talking about "different networks". While this could in theory be Lightning or sidechains, it's more likely to lead users toward scamcoins.

**Comment by aureleoules** [comment]
Thanks @luke-jr I addressed your suggestions.

File: src/key_io.cpp (hunk 3 above)
Comment (aureleoules) [reply src/key_io.cpp]: I also removed the `on %s` (network)

**Comment by willcl-ark** [comment]
Concept ACK

**Comment by davidgumberg** [comment]
Maybe add a message for not-yet-understood witness versions and a test case?

```diff
  if (version == 1 && data.size() == WITNESS_V1_TAPROOT_SIZE) {
      static_assert(WITNESS_V1_TAPROOT_SIZE == WitnessV1Taproot::size());
      WitnessV1Taproot tap;
      std::copy(data.begin(), data.end(), tap.begin());
      return tap;
  }
+
+ if (version > 1 && version <= 16) {
+     error_str =  "Invalid or unsupported Bech32 address witness version."
+     return CNoDestination();
+ }
+
  if (version > 16) {
      error_str = "Invalid Bech32 address witness version";
      return CNoDestination();
  }
```

**Comment by davidgumberg** [comment]
@sipa Thanks! I misunderstood.

**Review by ghost — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/26514/commits/962a0930e699b74b3c4d019427df6e2b3af5c831

---

> NACK on talking about "different networks". While this could in theory be Lightning or sidechains, it's more likely to lead users toward scamcoins.

Ideally it should have been 'chain' instead of network as networks are ipv4, ipv6, tor, i2p etc. although we should not be concerned about scammers if they ever use this software but use correct technical terms in errors IMO.

The comment is anyways addressed by author and I think PR improves error messages.

**Comment by davidgumberg** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/26514/commits/962a0930e699b74b3c4d019427df6e2b3af5c831

**Pull request merged**